packages/oid4vci-holder/src/agent/OID4VCIHolder.ts

```diff
@@ -0,0 +1,289 @@
+    oid4vciHolderAddContactIdentity: this.oid4vciHolderAddContactIdentity.bind(this),
+    oid4vciHolderAssertValidCredentials: this.oid4vciHolderAssertValidCredentials.bind(this),
+    oid4vciHolderStoreCredentialBranding: this.oid4vciHolderStoreCredentialBranding.bind(this),
+    oid4vciHolderStoreCredentials: this.oid4vciHolderStoreCredentials.bind(this),
+  }
+
+  private readonly vcFormatPreferences: Array<string> = ['jwt_vc_json', 'jwt_vc', 'ldp_vc']
+  private readonly onContactIdentityCreated?: (args: OnContactIdentityCreatedArgs) => Promise<void>
+  private readonly onCredentialStored?: (args: OnCredentialStoredArgs) => Promise<void>
+  private readonly onGetCredentials: (args: OnGetCredentialsArgs) => Promise<Array<CredentialToAccept>>
+
+  constructor(options: OID4VCIHolderOptions) {
+    const { onContactIdentityCreated, onCredentialStored, onGetCredentials, vcFormatPreferences } = options
+
+    if (vcFormatPreferences !== undefined && vcFormatPreferences.length > 0) {
+      this.vcFormatPreferences = vcFormatPreferences
+    }
+    this.onGetCredentials = onGetCredentials
+    this.onContactIdentityCreated = onContactIdentityCreated
+    this.onCredentialStored = onCredentialStored
+  }
+
+  public async onEvent(event: any, context: RequiredContext): Promise<void> {
+    switch (event.type) {
+      case OID4VCIHolderEvent.CONTACT_IDENTITY_CREATED:
+        this.onContactIdentityCreated?.(event.data)
+        break
+      case OID4VCIHolderEvent.CREDENTIAL_STORED:
+        this.onCredentialStored?.(event.data)
+        break
+      default:
+        return Promise.reject(Error('Event type not supported'))
+    }
+  }
+
+  private async oid4vciHolderGetMachineInterpreter(args: OID4VCIMachineInstanceOpts, context: RequiredContext): Promise<OID4VCIMachineInterpreter> {
+    const services = {
+      initiateOID4VCI: (args: InitiateOID4VCIArgs) => this.oid4vciHolderGetInitiationData(args, context),
+      createCredentialSelection: (args: CreateCredentialSelectionArgs) => this.oid4vciHolderCreateCredentialSelection(args, context),
+      getContact: (args: GetContactArgs) => this.oid4vciHolderGetContact(args, context),
+      getCredentials: (args: GetCredentialsArgs) => this.oid4vciHolderGetCredentials(args, context),
+      addContactIdentity: (args: AddContactIdentityArgs) => this.oid4vciHolderAddContactIdentity(args, context),
+      assertValidCredentials: (args: AssertValidCredentialsArgs) => this.oid4vciHolderAssertValidCredentials(args, context),
+      storeCredentialBranding: (args: StoreCredentialBrandingArgs) => this.oid4vciHolderStoreCredentialBranding(args, context),
+      storeCredentials: (args: StoreCredentialsArgs) => this.oid4vciHolderStoreCredentials(args, context),
+    }
+
+    const oid4vciMachineInstanceArgs: OID4VCIMachineInstanceOpts = {
+      ...args,
+      services: {
+        ...services,
+        ...args.services,
+      },
+    }
+
+    return OID4VCIMachine.newInstance(oid4vciMachineInstanceArgs)
```

Comment:
Could we also return a unique instance Id, or get that from the machine. So that we return an object. One being the interpreter and one being the id of the instance and for instance something like the name of the machine? That will be helpful probably for the (re)hydration support we are also creating

Reply:
refactored, there is no machine name. so no sure what you want to do here